abci/server/socket_server.go

```diff
@@ -154,14 +161,18 @@ func (s *SocketServer) handleRequests(closeConn chan error, conn net.Conn, respo
 			if err == io.EOF {
 				closeConn <- err
 			} else {
-				closeConn <- fmt.Errorf("Error reading message: %v", err.Error())
+				closeConn <- fmt.Errorf("error reading message: %v", err)
 			}
 			return
 		}
 		s.appMtx.Lock()
 		count++
 		s.handleRequest(req, responses)
 		s.appMtx.Unlock()
+
+		if panicVal != nil {
+			closeConn <- fmt.Errorf("recovered from panic: %v", panicVal)
```

Comment:
we also need to unlock the application mutex, no? `s.appMtx.Unlock()`

Reply:
AFAIK the execution is going to continue on the next line after the panic happened, so this would be `s.appMtx.Unlock()`, so by the time we are at this condition the mutex has already been unlocked.